url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/com/wroblicky/andrew/joust/pgn/PGNParser.java

```diff
@@ -0,0 +1,97 @@
+package com.wroblicky.andrew.joust.pgn;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.StringTokenizer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Converts the moves described in a PGN text file into a list of string tokens
+ *
+ * @author Andrew Wroblicky
+ *
+ */
+public final class PGNParser {
+
+	public static PGNGame getPGNGame(String pgnTextFile) {
+		PGNGame pgnGame = new PGNGame();
+		pgnGame.setFilename(new File(pgnTextFile).getName());
+		List<String> moves = new ArrayList<String>();
+		String next = null;
+
+		try {
+			BufferedReader in = new BufferedReader(new FileReader(pgnTextFile));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *in*. Execution paths that do not contain closure statements were detected. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).